test.js

```diff
@@ -1,51 +1,100 @@
 
   t.true(port.message.includes('should be >= 0 and < 65536. Received 65536.'));
 });
 
-test('should return `true` when available port is chosen', async (t) => {
+test.serial('should return `true` when available port is chosen', async (t) => {
   const port = await isPortFree(3007);
 
   t.is(port, true);
 });
 
-test('should return `false` when occupied is chosen', async (t) => {
-  const needfulPort = 3010;
+test.serial('should throw a port searching error when checking port outside the available range', async (t) => {
+  const port = await t.throwsAsync(() => isPortFree(65536));
+
+  t.true(port.message.includes('should be >= 0 and < 65536. Received 65536.'));
+});
+
+test.serial('should return the passed port value of the requested host', async (t) => {
+  const port = await findFreePort(3000, '127.0.0.1');
 
+  t.is(port, 3000);
+});
+
+test.serial('should return `false` when occupied port is chosen', wrapper, (t, done) => {
   const server = net.createServer();
-  await pify(server.listen.bind(server))(needfulPort, '0.0.0.0');
 
-  const port = await isPortFree(needfulPort);
+  server.once('listening', async () => {
+    server.once('close', () => setTimeout(done, 3000));
 
-  t.is(port, false);
+    const port = await isPortFree(requestedPort);
+    t.is(port, false);
+
+    server.close();
+  });
+
+  server.listen(requestedPort, defaultHost);
 });
 
-test('should throw a port searching error when checking port outside the available range', async (t) => {
-  const port = await t.throwsAsync(() => isPortFree(65536));
+test.serial('should return first available port', wrapper, (t, done) => {
+  const server = net.createServer();
 
-  t.true(port.message.includes('should be >= 0 and < 65536. Received 65536.'));
+  server.once('listening', async () => {
+    server.once('close', () => setTimeout(done, 3000));
+
+    const port = await findFreePort(requestedPort);
+    t.not(port, requestedPort);
+
+    server.close();
+  });
+
+  server.listen(requestedPort, defaultHost);
 });
 
-test('should return first available port', async (t) => {
-  const needfulPort = 3005;
+test.serial('should return `false` when occupied port of a requested host is chosen', wrapper, (t, done) => {
+  const server = net.createServer();
+
+  server.once('listening', async () => {
+    server.once('close', () => setTimeout(done, 3000));
 
+    const port = await isPortFree(requestedPort, requestedHost);
+    t.is(port, false);
+
+    server.close();
+  });
+
+  server.listen(requestedPort, requestedHost);
+});
+
+test.serial('should return `true` when port of requested host is free whereas the same port of the default host is occupied', wrapper, (t, done) => {
   const server = net.createServer();
-  await pify(server.listen.bind(server))(needfulPort, '0.0.0.0');
 
-  const port = await findFreePort(needfulPort);
+  server.once('listening', async () => {
+    server.once('close', () => setTimeout(done, 3000));
```

Comment:
So, i have noticed that when using the same port value in two tests, the second test will fail with an error:
`› process._tickCallback (internal/process/next_tick.js:63:19)`. 

The reason might be that once starting a server and not closing it we create conflicts for next tests. 
I have added listeners and closing events for those tests. Locally they pass, but not on CI for some reason.

Reply:
It looks like the reason of such weird behavior is the implementation of socket binding of Ubuntu instance that Github CI runs for our tests. I'm not sure what exactly is going wrong, but it looks like there's something “not normal” around `SO_REUSEADDR` feature.

I asked @mdnsk to run the tests on his Fedora, but they passed well. Then I asked @goganchic to run the test on his Ubuntu and suddenly this one failed.

Another one thought was “maybe Github CI runs some restricted virtualization layer and it fails everything”, but when I changed CI settings to run `macos-latest` instead of `ubuntu-latest` everything worked fine and tests passed.

So there's something strange with Ubuntu, but I'm too frontendish to find out the reason of such behavior. I have paws 🐾 

----

But what I know is the fact that we actually don't need this test suite. It doesn't test our library because it's relies on the OS or Node.js behavior which may be different (and that's exactly what we see here). The important part for us is that we can test a pair of `host:port` which we have bound, and the `host` value depends on the second param. As I see we've already tested such cases.

So let's just remove this suite ¯\_(ツ)_/¯.

Reply:
Thanks for researching this one. Deleted this OS dependent test.